mlir/lib/Conversion/VectorToSPIRV/VectorToSPIRV.cpp

```diff
@@ -381,9 +393,26 @@ struct VectorReductionPattern final
     if (Value acc = adaptor.getAcc())
       values.push_back(acc);
 
-    // Reduce them.
-    Value result = values.front();
-    for (Value next : llvm::ArrayRef(values).drop_front()) {
+    return values;
+  }
+};
+
+#define VECTOR_REDUCTION_BASE                                                  \
+  VectorReductionPatternBase<VectorReductionPattern<SPIRVUMaxOp, SPIRVUMinOp,  \
+                                                    SPIRVSMaxOp, SPIRVSMinOp>>
```

Comment:
Can we make this a `using` declaration instead?

Reply:
Unfortunately, no.
The macro is needed to avoid repetition between the base class and explicitly inheriting the constructor of the base class.

Reply:
There is other option though but it might look weird:
```cpp
template <typename SPIRVUMaxOp, typename SPIRVUMinOp, typename SPIRVSMaxOp,
          typename SPIRVSMinOp, typename..., 
          typename Base = VectorReductionPatternBase<VectorReductionPattern<SPIRVUMaxOp, SPIRVUMinOp, SPIRVSMaxOp, SPIRVSMinOp>>>
struct VectorReductionPattern final : Base {
  using Base::Base;
  ...
};
```
`typename...` is needed to protect from possible wrongful overrides of the `Base` and `typename Base` is used here as an alias.

Reply:
How about something like this:
```c++
template <typename... Args>
struct VectorReductionPattern;

typename <typename... Args>
using CrtpBase = VectorReductionPatternBase<VectorReductionPattern, Args...>;

template <typename... Args>
struct VectorReductionPattern final : CrtpBase<Args...> {
  using Base = CrtpBase<Args...>; 
   using Base::Base;
   // define SPIRVUMaxOp et al. in VectorReductionPatternBase
  ...
};
```

just an idea, I haven't tried to compile this

Reply:
Actually, it should work, I'll definitely give it a try! My idea wouldn't work because `VectorReductionPattern` is not declared before using it.